### PR TITLE
JS: parseJson tuning

### PR DESCRIPTION
**Pull request by ykuzmenko73**
1 000 000 cycles.

**master**:
  Chrome:  19153 ms
  FF:  23602 ms

**branch**:
  Chrome:  14065 ms
  FF: 22714 ms

Test App:

```
import formats/json/json;
import net/url_parameter;

main() {
	testJsons = [
		"{\"0\":{}}",
		"{\"1\": 1, \"2\": 2, \"3\": {\"4\": 4, \"5\": {\"6\": 6}}}",
		"{\"p\": 5}",
		"{\"0\":{\"0\":{\"0\":\"1\",\"1\":\"admin@area9.dk\",\"2\":\"Admin\",\"3\":\"Account\",\"4\":\"admin@area9.dk\",\"5\":\"\",\"6\":\"2016-03-15 13:48:28\",\"7\":\"1\",\"8\":\"\",\"9\":null,\"10\":null,\"11\":\"1\",\"12\":null,\"13\":\"0\"},\"1\":{\"0\":\"2\",\"1\":\"demo_old@area9\",\"2\":\"Old Demo\",\"3\":\"Account\",\"4\":\"demo_old@area9\",\"5\":\"\",\"6\":\"2016-03-15 13:48:28\",\"7\":\"1\",\"8\":\"\",\"9\":null,\"10\":null,\"11\":\"0\",\"12\":null,\"13\":\"0\"},\"2\":{\"0\":\"3\",\"1\":\"ulrik@area9.dk\",\"2\":\"Ulrik\",\"3\":\"Juul Christensen\",\"4\":\"ulrik@area9.dk\",\"5\":\"\",\"6\":\"2016-05-10 13:57:16\",\"7\":\"1\",\"8\":\"\",\"9\":null,\"10\":null,\"11\":\"1\",\"12\":null,\"13\":\"0\"},\"3\":{\"0\":\"4\",\"1\":\"yuriy.kuzmenko@area9.dk\",\"2\":\"Yuriy\",\"3\":\"Kuzmenko\",\"4\":\"yuriy.kuzmenko@area9.dk\",\"5\":\"\",\"6\":\"2019-01-25 12:08:07\",\"7\":\"1\",\"8\":\"\",\"9\":\"\",\"10\":null,\"11\":\"1\",\"12\":null,\"13\":\"0\"},\"4\":{\"0\":\"5\",\"1\":\"demo@area9.dk\",\"2\":\"Demo\",\"3\":\"Account\",\"4\":\"demo@area9.dk\",\"5\":\"\",\"6\":\"2019-02-26 09:43:26\",\"7\":\"1\",\"8\":\"\",\"9\":null,\"10\":null,\"11\":\"1\",\"12\":null,\"13\":\"0\"}},\"1\":{\"0\":{\"0\":\"1\"}},\"2\":{\"0\":{\"0\":\"1\"},\"1\":{\"0\":\"2\"},\"2\":{\"0\":\"3\"},\"3\":{\"0\":\"4\"},\"4\":{\"0\":\"5\"}},\"3\":{\"0\":{\"0\":\"1\"},\"1\":{\"0\":\"2\"},\"2\":{\"0\":\"3\"},\"3\":{\"0\":\"4\"},\"4\":{\"0\":\"5\"}},\"4\":{},\"5\":{\"0\":{\"0\":\"User\",\"1\":0},\"1\":{\"0\":\"can_create_User\",\"1\":1},\"2\":{\"0\":\"updatable_ids_User\",\"1\":2},\"3\":{\"0\":\"deletable_ids_User\",\"1\":3},\"4\":{\"0\":\"UsersSocial\",\"1\":4}}}",
	];

	cycles = s2i(getUrlParameterDef("cycles", "100000"));
	println("cycles: " + i2s(cycles));

	println("Test converting...");
	iter(testJsons, \json -> {
		println("Checking '"+json+"'...");
		parsed = parseJson(json);
		println("  -> " + toString(parsed));
	});
	println("  Done");

	println("Measuring...");
	t0 = timestamp();
	fori(1, cycles, \__ -> {
		iter(testJsons, \json -> {
			parseJson(json);
			{}
		});
	});
	println("  Done in " + i2s(round(timestamp() - t0)));
}
```

### CODE REVIEW

**Comment by ykuzmenko73** [comment]
[Investigate and tune SIMULATE for complex Adaptive Slide](https://trello.com/c/BfXjNEOs/736-investigate-and-tune-simulate-for-complex-adaptive-slide)

File: platforms/js/Native.hx [hunk 1]
```diff
@@ -1749,36 +1749,143 @@ class Native {
 	}
 
 	#if js
+	static var sidJsonArray;
```
Comment (alstrup) [inline platforms/js/Native.hx:1752]: Add a comment explaining what magic this is

Comment (alstrup) [reply platforms/js/Native.hx:1752]: Also give them type annotation to help clarify what is going on.

File: platforms/js/Native.hx [hunk 2]
```diff
@@ -1749,36 +1749,146 @@ class Native {
...
 	public static function parseJson(json : String) : Dynamic {
-	 	try {
-	 		var o = haxe.Json.parse(json);
-			return object2JsonStructs(o);
-		} catch (e : Dynamic) {
-			return makeStructValue("JsonDouble", [0.0], null); 
-		}
+		try {
+			if (parseJsonFirstCall) {
+				untyped Native.sidJsonArray = HaxeRuntime._structids_.h["JsonArray"];
```
Comment (VsevolodZakharov) [inline platforms/js/Native.hx:1865]: it should not rely on Haxe hashmap inner implementation and should use .get() here

File: platforms/js/Native.hx [hunk 3]
```diff
@@ -1749,36 +1749,146 @@ class Native {
...
+	// Chrome and maybe other browsers faster with for(var f in o) that with Object.getOwnPropertyNames
 	private static function object2JsonStructs(o : Dynamic) : Dynamic {
-		if (untyped __js__("Array.isArray(o)")) {
-			return HaxeRuntime.fastMakeStructValue("JsonArray", map(o, object2JsonStructs));
+		untyped __js__("
+		if (Array.isArray(o)) {
+			var a1 = Native.map(o,Native.object2JsonStructs);
+			var obj = { _id : Native.sidJsonArray};
+			obj[Native.sidJsonArrayVal] = a1;
+			return obj;
```
[the remark this answers is not in the document]
Comment (ykuzmenko73) [reply platforms/js/Native.hx:1777]: checked, ES5 way has the same speed in Chrome and 20% performance drop in FF

File: platforms/js/Native.hx [hunk 4]
```diff
@@ -1749,36 +1749,146 @@ class Native {
...
+					obj[Native.sidJsonStringVal] = o;
+					return obj;
+				case 'number':
+					var obj = { _id : Native.sidJsonDouble};
+					obj[Native.sidJsonDoubleVal] = o;
+					return obj;
+				case 'boolean':
+					var obj = { _id : Native.sidJsonBool};
+					obj[Native.sidJsonBoolVal] = o;
+					return obj;
```
[the remark this answers is not in the document]
Comment (ykuzmenko73) [reply platforms/js/Native.hx]: addressed

File: platforms/js/Native.hx (hunk 2 above)
Comment (ykuzmenko73) [reply platforms/js/Native.hx]: addressed

**Pull request merged**